src/tokens/multi/encodable.rs

```diff
@@ -0,0 +1,308 @@
+                            .owners
+                            .iter()
+                            .find_map(|(stored_owner, token_owner_data)| {
+                                (*stored_owner == owner).then_some(token_owner_data.balance)
+                            })
+                    })
+            },
+        )
+        .unwrap_or_default()
+    }
+
+    /// Returns an allowance of `owner`'s tokens for `operator`.
+    ///
+    /// - If `id` is [`Some`], firstly checks if `operator` is allowed for all
+    /// `owner`'s tokens, and if not, returns an approved amount of the tokens
+    /// with this `id`.
+    /// - If `operator` is allowed for all `owner`'s tokens, returns
+    /// [`Amount::MAX`], otherwise returns 0.
+    /// - If `id` is [`None`], only checks if `operator` is allowed for all
+    /// `owner`'s tokens.
+    pub fn allowance(&self, owner: Owner, operator: Operator, id: Option<&Id>) -> Amount {
+        let owner_search = |(_, token): &(_, Token)| {
+            token
+                .owners
+                .iter()
+                .find(|(stored_owner, _)| *stored_owner == owner)
+                .and_then(|(_, token_owner_data)| {
+                    token_owner_data
+                        .allowances
+                        .iter()
+                        .find_map(|(stored_operator, allowance)| {
+                            (*stored_operator == operator).then_some(*allowance)
+                        })
+                })
+        };
+
+        let id_search = || {
+            id.and_then(|unwrapped_id| {
+                self.tokens
+                    .iter()
+                    .find(|(stored_id, _)| stored_id == unwrapped_id)
+                    .and_then(owner_search)
+            })
+        };
+
+        self.owners
+            .iter()
+            .find(|(stored_owner, _)| *stored_owner == owner)
+            .and_then(|(_, general_owner_data)| {
+                general_owner_data
+                    .operators
+                    .contains(&operator)
+                    .then_some(Amount::MAX)
+                    .or_else(id_search)
+            })
+            .unwrap_or_default()
+    }
+
+    /// Gets an attribute with given `key` for the tokens with given `id`.
+    ///
+    /// Returns [`None`] if an attribute with given `key` doesn't exist.
+    ///
+    /// # Errors
+    /// - [`MTError::InsufficientAmount`] if there are no tokens with given
+    /// `id`.
+    pub fn get_attribute(&self, id: &Id, key: &Vec<u8>) -> Result<Option<&Vec<u8>>, MTError> {
+        self.tokens
+            .iter()
+            .find_map(|(stored_id, token)| (stored_id == id).then_some(&token.attributes))
+            .ok_or(MTError::InsufficientAmount)
+            .map(|attributes| {
+                attributes
+                    .iter()
+                    .find_map(|(stored_key, value)| (stored_key == key).then_some(value))
+            })
+    }
+}
+
+impl From<SuperMTState> for MTState {
```

Comment:
Shouldn't we support reverse conversion too? `impl From<MTState> for SuperMTState`

Reply:
I don't see any use cases for the reverse conversion. Perhaps @azoyan would find it useful for his state migration task?

Reply:
As for me, we could transform `CodeableMTState` to `MTState` after decoding (say after `msg::load`) and use it as a faster one (due to `HashMap` instead of `Vec`). But need feedback from users, we could add it later.